onnxruntime/core/providers/trt/trt_execution_provider.cc

```diff
@@ -0,0 +1,420 @@
+    output_info_[fused_node->Name()].push_back(output_binding_indexes);
+    output_info_[fused_node->Name()].push_back(output_dim_sizes);
+    output_shapes_[fused_node->Name()] = output_shapes;
+
+    // Create function state
+    NodeComputeInfo compute_info;
+    compute_info.create_state_func = [=](ComputeContext* context, FunctionState* state) {
+      std::unique_ptr<TRTFuncState> p = std::make_unique<TRTFuncState>();
+      *p = {context->allocate_func, context->release_func, context->allocator_handle, parsers_[context->node_name].get(), engines_[context->node_name].get(), contexts_[context->node_name].get(),
+            input_info_[context->node_name], output_info_[context->node_name], output_shapes_[context->node_name]};
+      *state = p.release();
+      return 0;
+    };
+
+    // Release function state
+    compute_info.release_state_func = [](FunctionState state) {
+      if (state)
+        delete static_cast<TRTFuncState*>(state);
+    };
+
+    // Create compute function
+    compute_info.compute_func = [](FunctionState state, ONNXRunTimeTensor* input_tensors, size_t num_inputs, ONNXRunTimeTensor* output_tensors, size_t num_outputs) {
+      ORT_UNUSED_PARAMETER(num_inputs);
+      ORT_UNUSED_PARAMETER(num_outputs);
+      TRTFuncState* trt_state = reinterpret_cast<TRTFuncState*>(state);
+      std::vector<int> input_indexes = (trt_state->input_info)[0];
+      std::vector<int> input_binding_indexes = (trt_state->input_info)[1];
+      std::vector<int> input_dim_sizes = (trt_state->input_info)[2];
+      std::vector<int> output_binding_indexes = (trt_state->output_info)[0];
+      std::vector<int> output_dim_sizes = (trt_state->output_info)[1];
+      std::vector<std::vector<int64_t>> output_shapes = trt_state->output_shapes;
+      int num_binding_inputs = input_binding_indexes.size();
+      int num_binding_outputs = output_binding_indexes.size();
+      cudaStream_t stream;
+      CHECK(cudaStreamCreate(&stream));
+      std::vector<void*> buffers(num_binding_inputs + num_binding_outputs);
+      std::vector<int> batch_size;
+
+      // Get batch size and allocate cuda memory for inputs
+      for (int i = 0, end = num_binding_inputs; i < end; ++i) {
+        const auto& tensor_input = input_tensors[input_indexes[i]];
+        const auto& tensor_shape = tensor_input.shape;
+        batch_size.push_back(tensor_shape[0]);
+
+        const float* input = static_cast<float*>(tensor_input.data);
+        CHECK(cudaMalloc(&buffers[input_binding_indexes[i]], batch_size[i] * input_dim_sizes[i] * sizeof(float)));
+        CHECK(cudaMemcpy(buffers[input_binding_indexes[i]], input, batch_size[i] * input_dim_sizes[i] * sizeof(float), cudaMemcpyHostToDevice));
+      }
+
+      // Allocate cuda memory for outputs
+      for (int i = 0, end = num_binding_outputs; i < end; ++i) {
+        CHECK(cudaMalloc(&buffers[output_binding_indexes[i]], batch_size[0] * output_dim_sizes[i] * sizeof(float)));
```

Comment:
>batch_size[0] [](start = 62, length = 13)

Should this be batch_size[i]? If so, can you add a unit test to make sure a batch size > 1 is tested?

Reply:
I don't quite understand the code though. batch_size[0] has the shape of an input doesn't it, but this code is to allocate outputs. 

---
In reply to: [262195033](https://github.com/Microsoft/onnxruntime/pull/382#discussion_r262195033) [](ancestors = 262195033)